Automation template detail [MAILPOET-4534] [MAILPOET-4537]

JanJakes opened the pull request:
## Description

Adds new automation template detail dialog + renders automation preview within it.

## Code review notes

Adding the automation editor store to render the automation preview is not the most elegant thing, but we're already using the same approach in automation analytics. I will consider creating a ticket to refactor this in some way.

## QA notes

Please test with premium on and off. The "<" and ">" buttons (and keyboard arrows) should skip premium templates when premium is not active, and coming soon templates always.

The same templates are also on the automation listing page (when you scroll down). Please check that too.

Please, also verify that the automation editor works, and no regressions were introduced.

## Linked PRs

_N/A_

## Linked tickets

[MAILPOET-4534]
[MAILPOET-4537]

## After-merge notes

_N/A_

## Tasks

- [x] I followed [best practices](https://codex.wordpress.org/I18n_for_WordPress_Developers) for translations
- [x] I added sufficient test coverage
- [x] I embraced TypeScript by either creating new files in TypeScript or converting existing JavaScript files when making changes


[MAILPOET-4534]: https://mailpoet.atlassian.net/browse/MAILPOET-4534?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[MAILPOET-4537]: https://mailpoet.atlassian.net/browse/MAILPOET-4537?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

JanJakes committed:
Implement automation template detail dialog

[MAILPOET-4534]

JanJakes committed:
Display error when creating automation from template failes

[MAILPOET-4534]

JanJakes committed:
Extract automation templates grid to a component

[MAILPOET-4534]

JanJakes committed:
Extract template detail dialog to templates grid

[MAILPOET-4534]

JanJakes committed:
Handle previous and next template buttons

[MAILPOET-4534]

JanJakes committed:
Add keyboard support for previous and next templates

[MAILPOET-4534]

JanJakes committed:
Add API endpoint to fetch automation template data

[MAILPOET-4537]

JanJakes committed:
Add automation context and registry to listing and templates pages

[MAILPOET-4537]

JanJakes committed:
Use @wordpress/hooks types in wp-js-hooks

[MAILPOET-4537]

JanJakes committed:
Implement automation template preview in template detail

[MAILPOET-4537]

JanJakes committed:
Add template preview loading spinner

[MAILPOET-4537]

JanJakes committed:
Fix editor centering for cases with custom separators

[MAILPOET-4537]

JanJakes committed:
Make template navigation circular

[MAILPOET-4534]

JanJakes committed:
Center automation flow in the preview

[MAILPOET-4537]

JanJakes committed:
Add icon to template detail error message

[MAILPOET-4534]

JanJakes committed:
Fix tests to work with the new automation template dialog

[MAILPOET-4534]

rodrigoprimo commented on mailpoet/assets/js/src/automation/templates/components/template-detail.tsx:
```diff
@@ -74,6 +74,14 @@ export function TemplateDetail({
             <Button icon={chevronRight} />
           </div>
           <div className="mailpoet-automation-template-detail-footer-actions">
+            {error && (
```
Should we add a way for the user to dismiss the error message or make it disappear after some time?

JanJakes replied:
Good idea, added: https://github.com/mailpoet/mailpoet/pull/5310/commits/41558f00d43c825a54dfac61436dc31419438510

rodrigoprimo commented on mailpoet/assets/js/src/automation/templates/components/template-detail.tsx:
```diff
@@ -0,0 +1,135 @@
+            <h1>{template.name}</h1>
+            {template.description}
+          </div>
+          <div className="mailpoet-automation-template-detail-preview">
```
Currently, if the user clicks multiple times on the left or right arrows, we are going to trigger a request to load the preview for several automations, but maybe we only need to load the last automation as, ultimately, that is the only one that will be displayed to the user? Also, the description of the automation loads faster than the preview, causing a weird behavior where the browser displays the description of the last automation, but we are still showing the preview of previous automations.

I don't think this is a blocker but more a suggestion for improvement.

JanJakes replied:
@rodrigoprimo
> Currently, if the user clicks multiple times on the left or right arrows, we are going to trigger a request to load the preview for several automations, but maybe we only need to load the last automation as, ultimately, that is the only one that will be displayed to the user?

I'm not too worried about a few extra requests in the admin, but this is still a problem because it will cause the arriving templates to "blink", and it may end up on a different one (race condition). Fixed with abort controller: https://github.com/mailpoet/mailpoet/pull/5310/commits/8764064d11ee3e25019889e5e690f0ee0af8ca72

> Also, the description of the automation loads faster than the preview, causing a weird behavior where the browser displays the description of the last automation, but we are still showing the preview of previous automations.

Do you see a preview of previous automations? There should be a spinner:

https://github.com/mailpoet/mailpoet/assets/141436/eccbbdbe-b0d7-4a6d-9e96-9673010a5931

rodrigoprimo replied:
Thanks for the fix.

rodrigoprimo commented on mailpoet/assets/js/src/automation/templates/components/template-detail.tsx:
```diff
@@ -0,0 +1,135 @@
+            <h1>{template.name}</h1>
+            {template.description}
+          </div>
+          <div className="mailpoet-automation-template-detail-preview">
```
I wonder if we should display the scrollbar in the preview only when one is needed?

JanJakes replied:
@rodrigoprimo Is it a horizontal, vertical, or both scrollbars? Could you please share a screenshot? On macOS there are no scrollbars, and I've only added styles with `overflow: auto` (none of them has `scroll`).

There is `overflow: scroll` on `.mailpoet-automation-editor-automation` for some time already... is that one perhaps causing the problem? Does switching to `auto` fix that for you? Thanks!

rodrigoprimo replied:
Here is a screenshot:

![Screenshot from 2023-12-01 11-41-14](https://github.com/mailpoet/mailpoet/assets/77215/131e61bd-7068-46fb-90ac-5715aa498086)

I can confirm that switching to `overflow: auto` on `.mailpoet-automation-editor-automation` fixes the issue.

rodrigoprimo commented on mailpoet/assets/js/src/automation/templates/components/template-detail.tsx:
```diff
@@ -0,0 +1,135 @@
+            <h1>{template.name}</h1>
+            {template.description}
+          </div>
+          <div className="mailpoet-automation-template-detail-preview">
```
If a given automation has a filter, clicking on the label "Filters" in the preview causes the site to load an empty page.

JanJakes replied:
I would consider that a bug, I think we can make it clickable. Fix: https://github.com/mailpoet/mailpoet/pull/5310/commits/c33d5626b9f62d02db8bb4d530c342b4faea4bca

rodrigoprimo replied:
The modal seems too small for the user to view its contents. I'm testing using Chrome. Do you get something different?

![Screenshot from 2023-12-01 11-35-30](https://github.com/mailpoet/mailpoet/assets/77215/d53af7d8-3cac-4908-9ea4-0948926c59b3)

In the browser console, I see:

```
react-dom.development.js:86 Warning: validateDOMNesting(...): <button> cannot appear as a descendant of <button>.
    at button
    at UnforwardedButton (webpack-internal:///../node_modules/.pnpm/@wordpress+components@23.5.0_@types+react@18.0.28_react-dom@18.2.0_react@18.2.0/node_modules/@wordpress/components/build-module/button/index.js:114:7)
    at FilterWrapper (webpack-internal:///./assets/js/src/automation/editor/components/filter-wrapper/index.tsx:28:5)
    at div
    at div
    at div
    at div
    at FiltersList (webpack-internal:///./assets/js/src/automation/editor/components/filters/list.tsx:36:5)
    at div
    at div
    at div
    at VisualElementHandler (webpack-internal:///../node_modules/.pnpm/framer-motion@7.10.3_react-dom@18.2.0_react@18.2.0/node_modules/framer-motion/dist/es/motion/utils/VisualElementHandler.mjs:8:1)
    at MotionComponent (webpack-internal:///../node_modules/.pnpm/framer-motion@7.10.3_react-dom@18.2.0_react@18.2.0/node_modules/framer-motion/dist/es/motion/index.mjs:50:65)
    at eval (webpack-internal:///../node_modules/.pnpm/@wordpress+components@23.5.0_@types+react@18.0.28_react-dom@18.2.0_react@18.2.0/node_modules/@wordpress/components/build-module/popover/index.js:86:12)
    at UnforwardedPopover (webpack-internal:///../node_modules/.pnpm/@wordpress+components@23.5.0_@types+react@18.0.28_react-dom@18.2.0_react@18.2.0/node_modules/@wordpress/components/build-module/popover/index.js:122:5)
    at div
    at UnforwardedDropdown (webpack-internal:///../node_modules/.pnpm/@wordpress+components@23.5.0_@types+react@18.0.28_react-dom@18.2.0_react@18.2.0/node_modules/@wordpress/components/build-module/dropdown/index.js:46:5)
    at FiltersChip (webpack-internal:///./assets/js/src/automation/editor/components/filters/chip.tsx:25:5)
    at div
    at button
    at Comp (webpack-internal:///../node_modules/.pnpm/reakit-system@0.15.2_react-dom@18.2.0_react@18.2.0/node_modules/reakit-system/es/createComponent.js:50:26)
    at div
    at Step (webpack-internal:///./assets/js/src/automation/editor/components/automation/step.tsx:57:5)
    at div
    at StepWrapper (webpack-internal:///./assets/js/src/automation/editor/step-wrapper.tsx:25:5)
    at FlowStep (webpack-internal:///./assets/js/src/automation/editor/components/automation/flow-step.tsx:26:56)
    at div
    at div
    at Flow (webpack-internal:///./assets/js/src/automation/editor/components/automation/flow.tsx:21:5)
    at div
    at div
    at div
    at Comp (webpack-internal:///../node_modules/.pnpm/reakit-system@0.15.2_react-dom@18.2.0_react@18.2.0/node_modules/reakit-system/es/createComponent.js:50:26)
    at Automation (webpack-internal:///./assets/js/src/automation/editor/components/automation/index.tsx:38:5)
    at TemplatePreview (webpack-internal:///./assets/js/src/automation/templates/components/template-preview.tsx:38:5)
    at div
    at div
    at div
    at div
    at div
    at StyleProvider (webpack-internal:///../node_modules/.pnpm/@wordpress+components@23.5.0_@types+react@18.0.28_react-dom@18.2.0_react@18.2.0/node_modules/@wordpress/components/build-module/style-provider/index.js:43:5)
    at div
    at UnforwardedModal (webpack-internal:///../node_modules/.pnpm/@wordpress+components@23.5.0_@types+react@18.0.28_react-dom@18.2.0_react@18.2.0/node_modules/@wordpress/components/build-module/modal/index.js:51:5)
    at eval (webpack-internal:///./assets/js/src/automation/templates/components/template-detail.tsx:87:5)
    at TemplatesGrid (webpack-internal:///./assets/js/src/automation/templates/components/templates-grid.tsx:37:5)
    at div
    at div
    at TabPanel (webpack-internal:///../node_modules/.pnpm/@wordpress+components@23.5.0_@types+react@18.0.28_react-dom@18.2.0_react@18.2.0/node_modules/@wordpress/components/build-module/tab-panel/index.js:90:5)
    at TabPanel
    at div
    at Templates
```

rodrigoprimo commented on mailpoet/assets/js/src/automation/templates/components/template-detail.tsx:
```diff
@@ -0,0 +1,135 @@
+            <h1>{template.name}</h1>
+            {template.description}
+          </div>
+          <div className="mailpoet-automation-template-detail-preview">
```
Since the user cannot interact with the automation preview, should we prevent it from changing the cursor when hovering over some of its elements?

JanJakes replied:
I think related to https://github.com/mailpoet/mailpoet/pull/5310#discussion_r1410957433, I'd leave it as it is.

rodrigoprimo reviewed (commented):
Good job! I left some questions for us to discuss.

JanJakes committed:
Abort outdated requests when fetching template previews

[MAILPOET-4534]

JanJakes committed:
Fix filters list to work in automation template previews

[MAILPOET-4534]

JanJakes committed:
Make create from template errors dismissible

[MAILPOET-4534]

rodrigoprimo commented:
As I'm switching teams, I won't be able to continue reviewing this. Could you please pick a new reviewer? Feel free to ping me if the new reviewer is not able to reproduce one of the pending issues.

JanJakes committed:
Add popover slot to automation to make popovers and dropdowns work everywhere

[MAILPOET-4534]

JanJakes committed:
Fix scrolling inside automation popovers

[MAILPOET-4534]

JanJakes committed:
Display scrollbars only when necessary

[MAILPOET-4534]

JanJakes committed:
Remove styles that are now covered by templates grid

[MAILPOET-4534]

JanJakes commented:
@brezocordero Would you mind taking a look at the last 3 commits? 🙏 It was only a few fixes based on Rodrigo's review.

brezocordero reviewed (dismissed):
Verified that the issues with scrolling and the filter popover are fixed.

Aschepikov commented:
Hi @JanJakes  When I am scrolling templates on the automation templates page before the image is downloaded I see this console error
![image](https://github.com/mailpoet/mailpoet/assets/113185437/1b5dbee0-dab5-463e-a919-da17b5ba1a9c)
If I will scroll it on the automations page I will see this console error and the unknown error occurred message on the page
![image](https://github.com/mailpoet/mailpoet/assets/113185437/1f1deab7-5c99-4491-953a-32bd8612b7b0)
Can you take a look please

JanJakes committed:
Do not report aborted requests as errors

[MAILPOET-4534]

JanJakes committed:
Catch request errors when loading template previews

[MAILPOET-4534]

JanJakes committed:
Add error message on preview request failure

[MAILPOET-4534]

JanJakes commented:
@Aschepikov Good catch, thanks!

Added a fix and an error state in case of real request error (not just aborting):

<img width="814" alt="Screenshot 2023-12-07 at 16 40 08" src="https://github.com/mailpoet/mailpoet/assets/141436/345303c9-aa40-40d6-be55-2c0d3f85fea7">

@brezocordero I added 3 more commits 🙏

brezocordero reviewed (approved):
Verified the last three commits handle the errors gracefully.

Aschepikov reviewed (approved):
GJ!
